### PR TITLE
Optimize expression variable data structures

### PR DESCRIPTION
**Pull request by m-kuhn**
Fewer code lines and a better datastructure :-)

Two QStringLists to one QVariantMap

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -7693,21 +7693,8 @@ void QgisApp::duplicateLayers( const QList<QgsMapLayer *>& lyrList )
...
-          {
-            break;
-          }
-
-          QVariant varValue = variableValues.at( varIndex );
-          varIndex++;
-          QgsExpressionContextUtils::setLayerVariable( dupVLayer, variableName, varValue );
-        }
+        QVariantMap vars = vlayer->customProperty( "expressionVariables" ).toMap();
+        dupVLayer->setCustomProperty( "expressionVariables", vars );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Shouldn't "duplicateLayers" copy ALL custom properties? This looks really fragile

Comment (m-kuhn) [inline src/app/qgisapp.cpp]: duplicateLayers tries what's possible to get a thing as similar as possible to the original. so why not...

**Comment by m-kuhn** [comment]
Incomplete, needs more work

**Comment by nyalldawson** [comment]
@m-kuhn should I be reviewing this?

**Comment by m-kuhn** [comment]
It's not urgent at all. I don't mind if it lands in 2.14 or not.

**Comment by nyalldawson** [comment]
I just wasn't sure if you were still working on it - it's marked as required changes.

**Comment by m-kuhn** [comment]
It does require changes. I want to continue it, but don't know when I'll find the time.

**Pull request closed without merging**